File: backend/mail/mail.go
package mail

import (
	"fmt"

	"github.com/sendgrid/sendgrid-go"
	"github.com/sendgrid/sendgrid-go/helpers/mail"
)

type Client struct {
	client    *sendgrid.Client
	baseEmail *mail.Email
}

type Request struct {
	Name string `json:"name"`
	Addr string `json:"addr"`
	Subj string `json:"subj"`
	Body string `json:"body"`
}

func NewClient(key, name, addr string) *Client {
	return &Client{
		client:    sendgrid.NewSendClient(key),
		baseEmail: mail.NewEmail(name, addr),
	}
}

func (c *Client) Send(mailReq *Request) error {
	from := mail.NewEmail(mailReq.Name, mailReq.Addr)
	subject := fmt.Sprintf("[%s] %s", mailReq.Name, mailReq.Subj)
	message := mail.NewSingleEmail(from, subject, c.baseEmail, "?", mailReq.Body)
	_, err := c.client.Send(message)
	return err
}

File: frontend/src/utils/mail/mail.ts
import getConfig from "next/config";
const {publicRuntimeConfig} = getConfig();

export interface MailRequest {
  name: string;
  email: string;
  subject: string;
  message: string;
};

export interface MailError {
  error: boolean;
  message: string;
};

export const sendMail: (req: MailRequest) => Promise<MailError> = async (req) => {
  const url: string = publicRuntimeConfig.BE + "/send-mail";
  const resp = await fetch(url, {
    method: 'POST', // *GET, POST, PUT, DELETE, etc.
    body: JSON.stringify(req) // body data type must match "Content-Type" header
  });
  let err: MailError = {error: false, message: ""};
  if (resp.status != 200) {
    err.error = true;
    err.message = await resp.text();
    return err;
  }
  return err;
}
